### ISSUE TITLE
DeleteOrphanFiles or ExpireSnapshots outofmemory

### ISSUE THREAD

**Issue by zhaoyunhai1216** [apache/iceberg#3703]
``
`   Column joinCond = nameEqual.and(actualContains);
     List<String> orphanFiles = actualFileDF.join(validFileDF, joinCond, "leftanti")
        .as(Encoders.STRING())
        .collectAsList();

    Tasks.foreach(orphanFiles)
        .noRetry()
        .suppressFailureWhenFinished()
        .onFailure((file, exc) -> LOG.warn("Failed to delete file: {}", file, exc))
        .run(deleteFunc::accept);`

spark all path collect to driver，When there are too many files，outofmemory ，It doesn't make sense， Deletion should be performed directly on the cluster rather than being collected to the driver for deletion，I can't carry it out now

**Comment by RussellSpitzer** [comment]
We have considered this but we were worried about overwhelming distributed file systems and losing control over the parallelism of the job. That said have you tried just increasing driver memory, we have folks doing some rather large remove orphan and expire snapshot jobs and they just use larger drivers when required (they also usually increase the parallelism of the driver delete process as well.)

**Comment by dchristle** [comment]
@RussellSpitzer We have also hit this issue after doing a large copy of rows into a single Iceberg table. We could have avoided it by more carefully partitioning before the insert, but it's too late for that. There are definitely too many files -- we've cleaned them up via calling `rewriteDataFiles` in a loop over subsets of the table. But we can't expire the snapshots in a similar piece-by-piece way. The driver reports that it cannot broadcast more than 8GB -- our driver memory is already set much higher than this, so increasing driver memory won't fix it.

Is there any way to partially distribute this operation to executors? For instance, in rough pseudo-code just based on copying the code from the original post (and making it more Scala-esque),
```
val joinCond = nameEqual.and(actualContains);
val orphanFiles = actualFileDF.join(validFileDF, joinCond, "leftanti").persist(DISK_ONLY)

val orphanFileCount = orphanFiles.count()

orphanFiles
    .repartition(Math.ceil(orphanFileCount/1000))
    .map{fileBatch => 
        Tasks.forEach(fileBatch)
        .noRetry()
        .suppressFailureWhenFinished()
        .onFailure((file, exec) => LOG.warn("Failed to delete file: {}", file, exec))
        .run(deleteFunc::accept)
}
```

Adding a small `sleep` call would limit the QPS against the distributed file store, if that's a concern.


But without a change like this, I'm not sure how we can expire old snapshots/delete orphan files. Are there any other workarounds, besides increasing driver memory?

**Comment by dchristle** [comment]
I'm following up to say I got `deleteOrphanFiles` to complete successfully. After bumping the memory, I was confused why I didn't see any output in the logs from an occasional `RetryHttpInitializer: Encountered status code 503 when sending DELETE request to URL` error. I let it run for more than 24 hours; it seemed like the driver was hung rather than deleting any orphan files. 

In other GitHub issues on deleting orphan files, increasing the number of threads is mentioned. I modified my Spark job to do this with `.executeDeleteWith`:

```
val executorService = Executors.newFixedThreadPool(30)

SparkActions
    .get()
    .deleteOrphanFiles(icebergTable)
    .executeDeleteWith(executorService)
    .execute()
```

The frequency of the 503 retry errors went up. My interpretation is these errors have some small fixed probability of occurring on a Google Storage delete operation. Since there are now 30 concurrent delete operations, the log message is seen more frequently.

I let this new job run for about 36 hours & it finished deleting orphan files successfully. I wonder if there's some way to emit periodic log messages indicating the number of files that have been deleted, perhaps every 5 minutes. Once my driver had sufficient memory, the deletes were likely happening correctly, but as a user, I was confused when I didn't see any log output. The delete orphan files operation is different from other maintenance operations -- it can't be seen in the Spark UI as a job or stage. 

Any thoughts on adding some periodic log outputs? @RussellSpitzer

**Comment by RussellSpitzer** [comment]
I don't think that would be unreasonable, although I'm about to write a quick patch for bulk deletes so things may change a bit.

**Comment by RLashofRegas** [comment]
@dchristle What was your solution that fixed the `Cannot broadcast the table that is larger than 8GB` issue? I just ran into the same problem on expire snapshots. I am using `max_concurrent_deletes => {concurrency}, stream_results => true` but this has made no difference.

**Comment by RLashofRegas** [comment]
@RussellSpitzer You mentioned you were working on a patch that might affect this. Should I expect the issue I mentioned above to go away if we upgraded to a more recent version of Iceberg? Currently on 0.14.

**Comment by RussellSpitzer** [comment]
Not for broadcast, for that you just need to disable broadcast join in spark

**Comment by RLashofRegas** [comment]
@RussellSpitzer Maybe I'm missing something but I don't think I can disable the broadcast in this case. `spark.sql.autoBroadcastJoinThreshold=-1` will only disable automatic broadcasts, but I can see in both [Iceberg 0.14 `buildValidContentFileWithTypeDF`](https://github.com/apache/iceberg/blob/6d2edd6284ebc5301dbe45376a31ca8316852a77/spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkAction.java#L129C26-L129C57) and [the equivalent function in main `contentFileDS`](https://github.com/apache/iceberg/blob/16c2af49a142d7044b619403de3e52aed13893e0/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkAction.java#L149) it looks like `sparkContext.broadcast(...)` is being called. If the table to be broadcast is larger than 8GB this will fail. Any ideas?

**Comment by RussellSpitzer** [comment]
That's the table metadata. If your JSON file is 8gb large you have other problems :)

**Comment by RLashofRegas** [comment]
Ah. Thanks! setting `spark.sql.autoBroadcastJoinThreshold=-1` fixed it.

**Comment by manuzhang** [comment]
@RussellSpitzer I wonder where we are now for this issue.

**Comment by jkolash** [comment]
I ran into this issue myself I will attempt to use 
```
spark.sql.autoBroadcastJoinThreshold=-1
```

**Comment by jkolash** [comment]
> I ran into this issue myself I will attempt to use
> 
> ```
> spark.sql.autoBroadcastJoinThreshold=-1
> ```

The approach I took was to take 1 pass without setting that flag then if there was a failure automatically re-try with that flag set.  That way I wouldn't be potentially introducing a performance regression when it wasn't necessary.

This could be done automatically in the iceberg project itself. I think this will repeatedly come up and would like to have expireSnapshots "just work" without extra glue code needed by users.

**Comment by IgorBerman** [comment]
we running maintenance jobs as part of long living spark context and see memory accumulating under memoryStore and it's related to ExpireSnapshotsSparkAction lambdas
see following screenshot from memory analyzer
![Image](https://github.com/user-attachments/assets/c0a459e6-aa45-4e50-978b-ab703b9e8aa9)
the objects that are stored there are related to SerializableTableWithSize object somehow
I don't understand yet how, but imo it's this::doExecute call inside 
```
    public ExpireSnapshots.Result execute() {
        JobGroupInfo info = this.newJobGroupInfo("EXPIRE-SNAPSHOTS", this.jobDesc());
        return (ExpireSnapshots.Result)this.withJobGroupInfo(info, this::doExecute);
    }
```

**Comment by IgorBerman** [comment]
It seems that after all it's related to broadcast that happens unconditionally in [contentFileDS](https://github.com/apache/iceberg/blob/1.8.x/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkAction.java#L153) method
the broadcast variable contains SerializableTableWithSize which is accumulated over time if same action is run in the same driver
there is no code that unpersists this broadcast.
I can't see a reason why setting spark.sql.autoBroadcastJoinThreshold=-1 should help here, since this is core spark api and not sql api

**Comment by IgorBerman** [comment]
ok, we found something in our environment. Disabling binPacking suddenly released this "leak"
what we think is that ContextCleaner component of spark that is responsible for cleanup of broadcasts is not keeping up with all the tasks it has. We see steady increase of referenceQueue length. 
Most probably most of the tasks there are of type cleanAcum, so those are kind of blocking broadcasts to be cleaned automatically. Imo it's best to cleanup those explicitly.
Meanwhile we are going to try spark.cleaner.referenceTracking.blocking  false and to reduce spark.cleaner.periodicGC.interval to some minimum possible , so that cleaner will kick in immediately.

**Comment by IgorBerman** [comment]
above settings helped to solve the issue for us
